File: LeetCodes/first-missing-positive/7-4-2023, 5_48_07 PM/first-missing-positive.cpp
// Link: https://leetcode.com/problems/first-missing-positive

/** Submission Info:
 *
 *  AUTHOR: the_247 | Raj Patel
 *  TITLE: first-missing-positive
 *  RUNTIME: 71 ms
 *  MEMORY: 41.1 MB
 *  DATE: 7-4-2023, 5:48:07 PM
 *
 */

// Solution:

class Solution {
public:
    int firstMissingPositive(vector<int>& nums) {
        int n = nums.size();
        for(int i=0; i<n; i++) {
            while(nums[i] > 0 && nums[i] <= n && nums[nums[i] - 1] != nums[i])
                swap(nums[i], nums[nums[i] - 1]);
        }
        for(int i=0; i<n; i++) {
            if(nums[i] != i+1)
                return i+1;
        }
        return n+1;
    }
};

File: LeetCodes/first-missing-positive/3-26-2024, 10_10_05 AM/first-missing-positive.java
// Link: https://leetcode.com/problems/first-missing-positive

/** Submission Info:
 *
 *  AUTHOR: the_247 | Raj Patel
 *  TITLE: first-missing-positive
 *  RUNTIME: 1 ms
 *  MEMORY: 55 MB
 *  DATE: 3-26-2024, 10:10:05 AM
 *
 */

// Solution:

class Solution {
    public int firstMissingPositive(int[] nums) {
        int n = nums.length;
        for(int i=0;i<n;i++) {
            if(nums[i] < 0) nums[i] = 0;
        }
        for(int i=0;i<n;i++) {
            int ind = Math.abs(nums[i]) - 1;
            if(ind < 0 || ind >= nums.length) 
                continue;

            if(nums[ind] > 0) {
                nums[ind] *= -1;
            } else if(nums[ind] == 0) {
                nums[ind] = Integer.MIN_VALUE + 1; 
            }
        }
        for(int i=0;i<n;i++) {
            if(nums[i] >= 0) return i+1;
        }
        return n+1;
    }
}
